File: Backend/soundwave/src/main/java/org/example/soundwave/service/ShippingMethodService.java
package org.example.soundwave.service;

import lombok.RequiredArgsConstructor;
import org.example.soundwave.model.dto.ShippingMethodDTO;
import org.example.soundwave.model.entity.ShippingMethod;
import org.example.soundwave.model.exception.ShippingMethodException;
import org.example.soundwave.model.request.ShippingMethodRequest;
import org.example.soundwave.model.response.PageResponse;
import org.example.soundwave.repository.ShippingMethodRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ShippingMethodService {
    private final ShippingMethodRepository shippingMethodRepository;

    public ShippingMethod findShippingMethodById(Long id){
        return shippingMethodRepository.findShippingMethodById(id)
                .orElseThrow(() -> new ShippingMethodException("Shipping method with id: " + id + " do not exist"));
    }

    public PageResponse<ShippingMethodDTO> getShippingMethods(int pageNo, int pageSize, String sortBy, String sortDir, String name) {
        Sort sort = sortDir.equalsIgnoreCase(Sort.Direction.ASC.name()) ?
                Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();

        Pageable pageable = PageRequest.of(pageNo, pageSize, sort);
        Page<ShippingMethod> shippingMethods = shippingMethodRepository.findByNameContainingIgnoreCase(name, pageable);

        List<ShippingMethodDTO> content = shippingMethods.getContent()
                .stream()
                .map(ShippingMethodDTO::new)
                .collect(Collectors.toList());

        return new PageResponse<>(
                content,
                shippingMethods.getNumber(),
                shippingMethods.getSize(),
                shippingMethods.getTotalElements(),
                shippingMethods.getTotalPages(),
                shippingMethods.isLast());
    }

    public List<ShippingMethodDTO> getAllShippingMethods() {
        return shippingMethodRepository.findAll()
                .stream()
                .sorted(Comparator.comparing(ShippingMethod::getPrice).reversed())
                .map(ShippingMethodDTO::new)
                .collect(Collectors.toList());
    }

    public void addShippingMethod(ShippingMethodRequest request) {
        ShippingMethod shippingMethod = ShippingMethod.builder()
                .name(request.name())
                .description(request.description())
                .price(request.price()).build();

        shippingMethodRepository.save(shippingMethod);
    }

    public void editShippingMethod(Long id, ShippingMethodRequest request) {
        ShippingMethod shippingMethod = findShippingMethodById(id);

        shippingMethod.setName(request.name());
        shippingMethod.setDescription(request.description());
        shippingMethod.setPrice(request.price());

        shippingMethodRepository.save(shippingMethod);
    }

    public void deleteShippingMethod(Long id) {
        ShippingMethod shippingMethod = findShippingMethodById(id);

        shippingMethodRepository.delete(shippingMethod);
    }
}

File: Frontend/soundwave/src/services/shippingMethodService.js
import { apiClient, handleApiError } from './apiService.js';

export const shippingMethodService = {
    async getShippingMethods(params = {}) {
        const {
            page = 0,
            size = 10,
            sortBy = 'id',
            sortDir = 'asc',
            name = ''
        } = params;

        const queryParams = new URLSearchParams();
        queryParams.append('page', page.toString());
        queryParams.append('size', size.toString());
        queryParams.append('sortBy', sortBy);
        queryParams.append('sortDir', sortDir);
        queryParams.append('name', name);

        try {
            const response = await apiClient.get(`/api/shipping_method?${queryParams.toString()}`);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to fetch shipping methods');
        }
    },

    async getAllShippingMethods() {
        try {
            const response = await apiClient.get('/api/shipping_method/all');
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to fetch shipping methods');
        }
    },

    async addShippingMethod(shippingMethodRequest) {
        try {
            const response = await apiClient.post('/api/shipping_method/add', shippingMethodRequest);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to add shipping method');
        }
    },

    async editShippingMethod(id, shippingMethodRequest) {
        try {
            const response = await apiClient.put(`/api/shipping_method/${id}`, shippingMethodRequest);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to edit shipping method');
        }
    },

    async deleteShippingMethod(id) {
        try {
            const response = await apiClient.delete(`/api/shipping_method/${id}`);
            return response.data;
        } catch (error) {
            handleApiError(error, 'Failed to delete shipping method');
        }
    }
};

export default shippingMethodService;